biodatasets/chemdner/chemdner.py

```diff
@@ -0,0 +1,392 @@
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, "development.bioc.xml"),
+                    "split": "dev",
+                },
+            ),
+        ]
+
+    def _get_passages_and_entities(self, d: bioc.BioCDocument) -> Tuple[List[Dict], List[List[Dict]]]:
+
+        passages: List[Dict] = []
+        entities: List[List[Dict]] = []
+
+        text_total_length = 0
+
+        po_start = 0
+
+        for _, p in enumerate(d.passages):
+
+            # annotation used only for document indexing
+            if p.text is None or len(p.text) <= 0:
+                continue
+
+            # annotation used only for document indexing
+            if p.annotations is None or len([a for a in p.annotations if a.text is not None and len(a.text) > 0]) <= 0:
+                continue
+
+            eo = p.offset - text_total_length
+
+            text_total_length += len(p.text) + 1
+
+            po_end = po_start + len(p.text)
+
+            dp = {
+                "text": p.text,
+                "type": p.infons.get("type"),
+                "offsets": [(po_start, po_end)],
+                "offset": p.offset,  # original offset
+            }
+
+            po_start = po_end + 1
+
+            passages.append(dp)
+
+            pe = []
+
+            for a in p.annotations:
+
+                if a.text is None or len(a.text) <= 0:
+                    continue
+
+                a_type = a.infons.get("class")
+                # a_type = a.infons.get("type")
+
+                # no in-text annotation: only for document indexing
+                if a_type in ["MeSH_Indexing_Chemical", "OTHER"]:
+                    continue
+
+                da = {
+                    "type": a_type,
+                    "offsets": [
+                        (
+                            loc.offset - eo,
+                            loc.offset + loc.length - eo,
+                        )
+                        for loc in a.locations
+                    ],
+                    "text": (a.text,),
+                    "id": a.id,
+                    "normalized": self._get_normalized(a),
+                }
+
+                pe.append(da)
+
+            entities.append(pe)
+
+        return passages, entities
+
+    def _get_normalized(self, a: bioc.BioCAnnotation) -> List[Dict]:
+        """
+        Get normalization DB and ID from annotation identifiers
+        """
+
+        identifiers = a.infons.get("identifier")
+
+        if identifiers is not None:
+
+            identifiers = re.split(r",|;", identifiers)
+
+            identifiers = [i for i in identifiers if i != "-"]
+
+            normalized = [i.split(":") for i in identifiers]
+
+            normalized = [{"db_name": elems[0], "db_id": elems[1]} for elems in normalized]
+
+        else:
+
+            normalized = [{"db_name": "-1", "db_id": "-1"}]

```

Comment:
Are missing normalisations handled with -1 also in the original data? If so, it's fine, otherwise I would replace it with `normalized = []`

Reply:
I will replace it to `normalized = []` since no normalization is specified.